### PR TITLE
[RF] Added code generation support for RooEfiiciency

### PR DESCRIPTION
**Pull request by abhigyan-ach**
# This Pull request: Added codegen support for RooEfficiency

### CODE REVIEW

File: roofit/roofit/inc/RooHistConstraint.h [hunk 1]
```diff
@@ -1,8 +1,12 @@
-/*****************************************************************************
- * Project: RooFit                                                           *
- *                                                                           *
-  * This code was autogenerated by RooClassFactory                            *
- *****************************************************************************/
+/*
+ * Project: RooFit
+ *
+ * Copyright (c) 2023, CERN
```
[the remark this answers is not in the document]
Comment (abhigyan-ach) [reply roofit/roofit/inc/RooHistConstraint.h:4]: That's weird I didn't make any changes to that file at all.

**Comment by guitargeek** [comment]
@phsft-bot build

File: roofit/roofitcore/inc/RooFit/Detail/EvaluateFuncs.h [hunk 2]
```diff
@@ -18,7 +18,27 @@
...
+#include <functional>
+#include <cassert>
+#include <array>
+#include <cmath>
+#include <span>
+#include <utility>
+#include "Riostream.h"
+
+#include "TClass.h"
+#include "RooRombergIntegrator.h"
```
Comment (vgvassilev) [inline roofit/roofitcore/inc/RooFit/Detail/EvaluateFuncs.h:33]:
```suggestion
```

File: roofit/roofitcore/inc/RooFit/Detail/EvaluateFuncs.h [hunk 3]
```diff
@@ -41,6 +41,17 @@ inline double bifurGaussEvaluate(double x, double mean, double sigmaL, double si
    return gaussianEvaluate(x, mean, sigmaR);
 }
 
+inline double RooEfficiencyEvaluate(double effFuncVal, int catIndex, int sigCatIndex)
```
Comment (vgvassilev) [inline roofit/roofitcore/inc/RooFit/Detail/EvaluateFuncs.h:44]:
```suggestion
inline double efficiencyEvaluate(double effFuncVal, int catIndex, int sigCatIndex)
```

File: roofit/roofitcore/src/RooEfficiency.cxx [hunk 4]
```diff
@@ -111,6 +111,11 @@ double RooEfficiency::analyticalIntegral(Int_t code, const char* /*rangeName*/)
   return 1.0 ;
 }
 
+void RooEfficiency::translate(RooFit::Detail::CodeSquashContext &ctx) const
+{
+    int sigCatIndex = _cat.arg().lookupIndex(_sigCatName.Data());
+    ctx.addResult(this, ctx.buildCall("RooFit::Detail::EvaluateFuncs::RooEfficiencyEvaluate",  _effFunc, _cat, sigCatIndex));
```
Comment (vgvassilev) [inline roofit/roofitcore/src/RooEfficiency.cxx:117]:
```suggestion
    ctx.addResult(this, ctx.buildCall("RooFit::Detail::EvaluateFuncs::efficiencyEvaluate",  _effFunc, _cat, sigCatIndex));
```

@guitargeek, I realize that `RooEfficiency::evaluate` is quite different. What can we do to reduce the code duplication here?

**Comment by vgvassilev** [comment]
@phsft-bot build!

File: roofit/roofitcore/src/RooEfficiency.cxx (hunk 4 above)
Comment (guitargeek) [reply roofit/roofitcore/src/RooEfficiency.cxx:117]: It was mathematically not different, and I updated the PR to use the new `efficiencyEvaluate` also in `RooEfficiency::evaluate`.

**Comment by guitargeek** [comment]
@phsft-bot build

**Review by guitargeek — APPROVED** [review]
Thank you very much for implementing this! I just addressed the comments of Vassil.

**Comment by guitargeek** [comment]
@phsft-bot build

**Pull request merged**